ext/sodium/libsodium.c

```diff
@@ -0,0 +1,2682 @@
+		zend_throw_exception(sodium_exception_ce, "crypto_generichash(): unsupported key length", 0);
+		return;
+	}
+	hash = zend_string_alloc(hash_len, 0);
+	if (crypto_generichash((unsigned char *) ZSTR_VAL(hash), (size_t) hash_len,
+						   msg, (unsigned long long) msg_len,
+						   key, (size_t) key_len) != 0) {
+		zend_string_free(hash);
+		zend_throw_exception(sodium_exception_ce, "crypto_generichash(): internal error", 0);
+		return;
+	}
+	ZSTR_VAL(hash)[hash_len] = 0;
+
+	RETURN_STR(hash);
+}
+
+PHP_FUNCTION(sodium_crypto_generichash_init)
+{
+	crypto_generichash_state  state_tmp;
+	zend_string				 *state;
+	unsigned char			 *key = NULL;
+	size_t					  state_len = sizeof (crypto_generichash_state);
+	zend_long				  hash_len = crypto_generichash_BYTES;
+	size_t					  key_len = 0;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "|sl",
+							  &key, &key_len,
+							  &hash_len) == FAILURE) {
+		return;
+	}
+	if (hash_len < crypto_generichash_BYTES_MIN ||
+		hash_len > crypto_generichash_BYTES_MAX) {
+		zend_throw_exception(sodium_exception_ce, "crypto_generichash_init(): unsupported output length", 0);
+		return;
+	}
+	if (key_len != 0 &&
+		(key_len < crypto_generichash_KEYBYTES_MIN ||
+		 key_len > crypto_generichash_KEYBYTES_MAX)) {
+		zend_throw_exception(sodium_exception_ce, "crypto_generichash_init(): unsupported key length", 0);
+		return;
+	}
+	if (crypto_generichash_init((void *) &state_tmp, key, (size_t) key_len,
+								(size_t) hash_len) != 0) {
+		zend_throw_exception(sodium_exception_ce, "crypto_generichash_init(): internal error", 0);
+		return;
+	}
+	state = zend_string_alloc(state_len, 0);
+	memcpy(ZSTR_VAL(state), &state_tmp, state_len);
+	sodium_memzero(&state_tmp, sizeof state_tmp);
+	ZSTR_VAL(state)[state_len] = 0;
+
+	RETURN_STR(state);
+}
+
+PHP_FUNCTION(sodium_crypto_generichash_update)
+{
+	crypto_generichash_state  state_tmp;
+	zval					 *state_zv;
+	unsigned char			 *msg;
+	unsigned char			 *state;
+	size_t					  msg_len;
+	size_t					  state_len;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "zs",
+							  &state_zv, &msg, &msg_len) == FAILURE) {
+		return;
+	}
+	ZVAL_DEREF(state_zv);
+	if (Z_TYPE_P(state_zv) != IS_STRING) {
+		zend_throw_exception(sodium_exception_ce, "crypto_generichash_update: a reference to a state is required", 0);
+		return;
+	}
+	sodium_separate_string(state_zv);
+	state = (unsigned char *) Z_STRVAL(*state_zv);
+	state_len = Z_STRLEN(*state_zv);
+	if (state_len != sizeof (crypto_generichash_state)) {
+		zend_throw_exception(sodium_exception_ce, "crypto_generichash_update(): incorrect state length", 0);
+		return;
+	}
+	memcpy(&state_tmp, state, sizeof state_tmp);
+	if (crypto_generichash_update((void *) &state_tmp, msg,
+								  (unsigned long long) msg_len) != 0) {
+		zend_throw_exception(sodium_exception_ce, "crypto_generichash_update(): internal error", 0);
+		return;
+	}
+	memcpy(state, &state_tmp, state_len);
+	sodium_memzero(&state_tmp, sizeof state_tmp);
+
+	RETURN_TRUE;
```

Comment:
This function throws an exception on error and returns `true` on success? Why is it not simply `void`? The returned value is meaningless because the user has to surround it with a try–catch anyways.